torch/lib/THC/generic/THCTensorRandom.cu

```diff
@@ -437,12 +437,20 @@ DEFINE_BERNOULLI_TENSOR(bernoulli_FloatTensor, THCudaTensor, float)
 DEFINE_BERNOULLI_TENSOR(bernoulli_DoubleTensor, THCudaDoubleTensor, double)
 
 #if defined(THC_REAL_IS_DOUBLE)
-
 GENERATE_KERNEL1(generate_geometric, double, double p, double, curand_uniform_double, ceil(log(x) / log(1-p)))
 #else
 GENERATE_KERNEL1(generate_geometric, real, double p, float, curand_uniform, (ScalarConvert<float, real>::to(ceilf(logf(x) / log(1-p)))))
 #endif
 
+#define CURAND64(STATE) (((uint64_t)curand(&state[blockIdx.x])) << 32) + (uint64_t)curand(&state[blockIdx.x])
+#if defined(THC_REAL_IS_DOUBLE) || defined(THC_REAL_IS_LONG)
+GENERATE_KERNEL2(generate_random, real, int64_t base, uint64_t range, uint64_t, CURAND64, (real)(x % range + base))
+#elif defined(THC_REAL_IS_FLOAT) || defined(THC_REAL_IS_HALF)
+GENERATE_KERNEL2(generate_random, real, int32_t base, uint32_t range, uint32_t, curand, (ScalarConvert<double, real>::to(double(x % range + base))))
```

Comment:
It's weird that for float it's going to be `long -> double -> float` while you could have just casted it to float directly